README.md
# restbl

A simple library to handle RSTB/RESTBL (resource size table) files from *The
Legend of Zelda: Tears of the Kingdom*. Features:
- Quick, zero-allocation parser
- Optional `alloc` feature to support editable table which can be serialized to
  binary or (with the `yaml` feature) YAML.
- `no_std` support (optional `std` feature)
- optional Serde support (`serde` feature)
- `aarch64-nintendo-switch-freestanding` support (without the `std` feature)

## Example Usage

```rust
use restbl::bin::ResTblReader;

let bytes = std::fs::read("test/ResourceSizeTable.Product.110.rsizetable").unwrap();

// Setup the quick, zero-allocation reader
let reader = ResTblReader::new(bytes.as_slice()).unwrap();
// Lookup an RSTB value
assert_eq!(
    reader.get("Bake/Scene/MainField_G_26_43.bkres"),
    Some(31880)
);

#[cfg(feature = "alloc")]
{
    use restbl::ResourceSizeTable;
    // Parse RSTB into owned table
    let mut table = ResourceSizeTable::from_parser(&reader);
    // Set the size for a resource
    table.set("TexToGo/Etc_BaseCampWallWood_A_Alb.txtg", 777);
    // Check the size
    assert_eq!(
        table.get("TexToGo/Etc_BaseCampWallWood_A_Alb.txtg"),
        Some(777)
    );
    // Dump to YAML, if `yaml` feature enabled
    #[cfg(feature = "yaml")]
    {
        let json_table = table.to_text();
        // From YAML back to RSTB
        let new_table = ResourceSizeTable::from_text(&json_table).unwrap();
    }
}
```

## Building for Switch

To build for Switch, you will need to use the
`aarch64-nintendo-switch-freestanding` target. The `std` feature is not
supported, so you will need to use `--no-default-features`. Since [`cargo
nx`](https://github.com/aarch64-switch-rs/cargo-nx) does not seem to support
passing feature flags, you will need to run the full command yourself, as
follows:

```
cargo build -Z build-std=core,compiler_builtins,alloc --target aarch64-nintendo-switch-freestanding --no-default-features
```

## License

This software is licensed under the terms of the GNU General Public License,
version 3 or later.

License: GPL-3.0-or-later


CHANGELOG.md
# Changelog

All notable changes to restbl will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.1.0]

- Initial release, mostly full-featured except for hash-decoding


src/bin.rs
#![allow(clippy::unused_unit)]
#![cfg_attr(not(feature = "alloc"), allow(clippy::needless_borrow))]
#[cfg(feature = "alloc")]
use alloc::borrow::Cow;
use core::mem::size_of;
use memoffset::offset_of;
use sa::static_assert;

use crate::{
    util::{hash_name, read_u32, Name},
    Error, Result, TableIndex,
};

/// Constant representing the magic of an RESTBL file
pub const MAGIC: &[u8] = b"RESTBL";

/// Represents an RESTBL header, without the magic
#[repr(C)]
#[derive(Debug, Clone, Copy)]
pub struct Header {
    version: u32,
    string_block_size: u32,
    crc_table_count: u32,
    name_table_count: u32,
}
static_assert!(Header::FULL_SIZE == 0x16);

impl Header {
    const FULL_SIZE: usize = size_of::<Header>() + MAGIC.len();

    #[inline(always)]
    pub fn version(&self) -> u32 {
        self.version
    }

    #[inline(always)]
    pub fn string_block_size(&self) -> u32 {
        self.string_block_size
    }

    #[inline(always)]
    pub fn crc_table_count(&self) -> u32 {
        self.crc_table_count
    }

    #[inline(always)]
    pub fn name_table_count(&self) -> u32 {
        self.name_table_count
    }

    /// Attempt to parse the RESTBL header from a slice
    fn read(data: &[u8]) -> Result<Self> {
        if data.len() < Self::FULL_SIZE {
            Err(Error::InsufficientData(data.len(), "0x16 bytes for header"))
        } else if &data[..MAGIC.len()] != MAGIC {
            Err(Error::InvalidMagic(
                data[..MAGIC.len()]
                    .try_into()
                    .expect("Slice must be 6 bytes long"),
            ))
        } else {
            let data = &data[MAGIC.len()..Self::FULL_SIZE];
            Ok(Self {
                version: read_u32(data, None)?,
                string_block_size: read_u32(data, Some(offset_of!(Header, string_block_size)))?,
                crc_table_count: read_u32(data, Some(offset_of!(Header, crc_table_count)))?,
                name_table_count: read_u32(data, Some(offset_of!(Header, name_table_count)))?,
            })
        }
    }

    /// Attempt to serialize the RESTBL header to a writer
    pub fn write(self, buffer: &mut [u8]) {
        buffer[..MAGIC.len()].copy_from_slice(MAGIC);
        buffer[MAGIC.len()..Self::FULL_SIZE].copy_from_slice(
            unsafe { core::mem::transmute::<Self, [u8; core::mem::size_of::<Self>()]>(self) }
                .as_slice(),
        )
    }
}

/// Represents a RESTBL hash entry
#[repr(C)]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct HashEntry {
    hash: u32,
    value: u32,
}
static_assert!(size_of::<HashEntry>() == 0x8);

impl HashEntry {
    /// Attempt to parse a RESTBL hash entry from a slice
    pub fn read(buffer: &[u8]) -> Result<Self> {
        if buffer.len() < size_of::<HashEntry>() {
            Err(Error::InsufficientData(
                buffer.len(),
                "8 bytes for HashEntry",
            ))
        } else {
            Ok(Self {
                hash: read_u32(buffer, None)?,
                value: read_u32(buffer, Some(offset_of!(HashEntry, value)))?,
            })
        }
    }

    /// Serialize a RESTBL hash entry to a buffer
    pub fn write(self, buffer: &mut [u8]) {
        buffer[..size_of::<Self>()].copy_from_slice(
            unsafe { core::mem::transmute::<Self, [u8; core::mem::size_of::<Self>()]>(self) }
                .as_slice(),
        );
    }

    #[inline(always)]
    pub fn hash(&self) -> u32 {
        self.hash
    }

    #[inline(always)]
    pub fn value(&self) -> u32 {
        self.value
    }
}

/// Represents a RESTBL name entry
#[repr(C)]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct NameEntry {
    name: Name,
    value: u32,
}
static_assert!(size_of::<NameEntry>() == 0xa4);

impl NameEntry {
    /// Attempt to parse a RESTBL name entry from a slice
    pub fn read(buffer: &[u8]) -> Result<Self> {
        if buffer.len() < size_of::<NameEntry>() {
            Err(Error::InsufficientData(
                buffer.len(),
                "0x4a bytes for NameEntry",
            ))
        } else {
            Ok(Self {
                name: Name::try_from(&buffer[..160])?,
                value: read_u32(buffer, Some(160))?,
            })
        }
    }

    /// Serialize a RESTBL name entry to a buffer
    pub fn write(self, buffer: &mut [u8]) {
        buffer[..size_of::<Self>()].copy_from_slice(
            unsafe { core::mem::transmute::<Self, [u8; core::mem::size_of::<Self>()]>(self) }
                .as_slice(),
        );
    }

    #[inline(always)]
    pub fn name(&self) -> Name {
        self.name
    }

    #[inline(always)]
    pub fn value(&self) -> u32 {
        self.value
    }
}

/// An extremely fast, non-allocating reader to parse an RSTB file
pub struct ResTblReader<'a> {
    #[cfg(feature = "alloc")]
    data: Cow<'a, [u8]>,
    #[cfg(not(feature = "alloc"))]
    data: &'a [u8],
    header: Header,
}

#[derive(Debug, Copy, Clone, PartialEq, Eq)]
pub enum TableEntry {
    Hash(HashEntry),
    Name(NameEntry),
}

struct HashTableIndex(usize);
struct NameTableIndex(usize);

/// Struct to iterate over RSTB entries
pub struct ResTblIterator<'a> {
    table: &'a ResTblReader<'a>,
    index: usize,
}

impl<'a> Iterator for ResTblIterator<'a> {
    type Item = TableEntry;

    fn next(&mut self) -> Option<Self::Item> {
        if let Some(hash_index) = self.table.hash_table_index(self.index) {
            let entry = self.table.parse_hash_entry(hash_index);
            self.index += 1;
            Some(TableEntry::Hash(entry))
        } else {
            let start = self.table.name_table_offset()
                + (self.index - self.table.header.crc_table_count as usize)
                    * size_of::<NameEntry>();
            let end = start + size_of::<NameEntry>();
            if end > self.table.data.len() {
                None
            } else {
                let data = &self.table.data[start..start + size_of::<NameEntry>()];
                let entry = NameEntry::read(data).ok();
                self.index += 1;
                entry.map(TableEntry::Name)
            }
        }
    }
}

#[cfg(feature = "alloc")]
type Buffer<'a> = alloc::borrow::Cow<'a, [u8]>;
#[cfg(not(feature = "alloc"))]
type Buffer<'a> = &'a [u8];

impl<'a> ResTblReader<'a> {
    /// Construct a new RSTB parser
    pub fn new<D: Into<Buffer<'a>>>(data: D) -> Result<Self> {
        fn inner(data: Buffer<'_>) -> Result<ResTblReader<'_>> {
            let header = Header::read(&data[..Header::FULL_SIZE])?;
            let expected_size = Header::FULL_SIZE
                + header.crc_table_count as usize * size_of::<HashEntry>()
                + header.name_table_count as usize * size_of::<NameEntry>();
            if data.len() < expected_size {
                Err(Error::InvalidTableSize(data.len(), expected_size))
            } else {
                Ok(ResTblReader { data, header })
            }
        }
        inner(data.into())
    }

    #[inline(always)]
    pub fn header(&self) -> &Header {
        &self.header
    }

    /// Get the total number of hash and name entries in the table
    #[inline(always)]
    pub fn len(&self) -> usize {
        (self.header.crc_table_count + self.header.name_table_count) as usize
    }

    #[inline(always)]
    pub fn is_empty(&self) -> bool {
        self.len() == 0
    }

    /// SAFETY: This involves two unsafe operations, `core::mem::transmute` and
    /// unchecked slice-to-array. They are perfectly sound, however. The slice
    /// conversion is sound because the size of the slice and the size of the
    /// array are both set specifically by the size of `HashEntry`, the slice is
    /// guaranteed to be within bounds because the table size was checked in the
    /// `new()` method, the only way to construct this parser, and the index
    /// type is only ever constructed if within the bounds of the hash table. The
    /// transmute is sound because any possible combination of 8 bytes can be
    /// legitimately interpreted as a pair of `u32` values. The values could be
    /// nonsense if the file is not valid, but they cannot produce undefined
    /// behavior.
    fn parse_hash_entry(&self, index: HashTableIndex) -> HashEntry {
        debug_assert!(index.0 < self.header.crc_table_count as usize);
        let start = Header::FULL_SIZE + index.0 * size_of::<HashEntry>();
        let end = start + size_of::<HashEntry>();
        unsafe {
            core::mem::transmute::<[u8; size_of::<HashEntry>()], HashEntry>(
                self.data
                    .get_unchecked(start..end)
                    .try_into()
                    .unwrap_unchecked(),
            )
        }
    }

    /// Check if the specified hash or resource name is present in the table.
    /// Checks the name table first (if applicable) and then the hash table.
    pub fn contains<'i, I: Into<TableIndex<'i>>>(&self, needle: I) -> bool {
        fn inner(tbl: &ResTblReader, needle: TableIndex) -> bool {
            match needle {
                TableIndex::HashIndex(hash) => tbl.find_hash_entry(hash).is_some(),
                TableIndex::StringIndex(name) => {
                    tbl.find_name_entry(&name).map(|e| e.value).is_some() || {
                        let hash = hash_name(&name);
                        tbl.find_hash_entry(hash).map(|e| e.value).is_some()
                    }
                }
            }
        }
        inner(self, needle.into())
    }

    /// Returns the RSTB value for the specified hash or resource name if present.
    /// Checks the name table first (if applicable) and then the hash table.
    pub fn get<'i, I: Into<TableIndex<'i>>>(&self, needle: I) -> Option<u32> {
        fn inner(tbl: &ResTblReader, needle: TableIndex) -> Option<u32> {
            match needle {
                TableIndex::HashIndex(hash) => tbl.find_hash_entry(hash).map(|e| e.value),
                TableIndex::StringIndex(name) => {
                    tbl.find_name_entry(&name).map(|e| e.value).or_else(|| {
                        let hash = hash_name(&name);
                        tbl.find_hash_entry(hash).map(|e| e.value)
                    })
                }
            }
        }
        inner(self, needle.into())
    }

    fn parse_name_entry(&self, index: NameTableIndex) -> Result<NameEntry> {
        let start = self.name_table_offset() + index.0 * size_of::<NameEntry>();
        let end = start + size_of::<NameEntry>();
        NameEntry::read(&self.data[start..end])
    }

    fn find_hash_entry(&self, hash: u32) -> Option<HashEntry> {
        let mut start = 0;
        let mut end = self.header.crc_table_count as usize;
        while start < end {
            let mid = (start + end) / 2;
            let entry = self.parse_hash_entry(HashTableIndex(mid));
            match entry.hash.cmp(&hash) {
                core::cmp::Ordering::Less => {
                    start = mid + 1;
                }
                core::cmp::Ordering::Greater => {
                    end = mid;
                }
                core::cmp::Ordering::Equal => return Some(entry),
            }
        }
        None
    }

    fn find_name_entry(&self, name: &str) -> Option<NameEntry> {
        let mut start = 0;
        let mut end = self.header.name_table_count as usize;
        while start < end {
            let mid = (start + end) / 2;
            let entry = self.parse_name_entry(NameTableIndex(mid)).ok()?;
            match entry.name.partial_cmp(&name) {
                Some(core::cmp::Ordering::Less) => {
                    start = mid + 1;
                }
                Some(core::cmp::Ordering::Greater) => {
                    end = mid;
                }
                Some(core::cmp::Ordering::Equal) => return Some(entry),
                _ => return None,
            }
        }
        None
    }

    /// Returns the RSTB entry for the specified hash or resource name if present.
    /// Checks the name table first (if applicable) and then the hash table.
    pub fn get_entry<'i, I: Into<TableIndex<'i>>>(&self, needle: I) -> Option<TableEntry> {
        fn inner(tbl: &ResTblReader, needle: TableIndex) -> Option<TableEntry> {
            match needle {
                TableIndex::HashIndex(hash) => tbl.find_hash_entry(hash).map(TableEntry::Hash),
                TableIndex::StringIndex(name) => tbl
                    .find_name_entry(&name)
                    .map(TableEntry::Name)
                    .or_else(|| {
                        let hash = hash_name(&name);
                        tbl.find_hash_entry(hash).map(TableEntry::Hash)
                    }),
            }
        }
        inner(self, needle.into())
    }

    /// Iterate all RSTB entries across both the hash and name tables.
    pub fn iter(&self) -> ResTblIterator<'_> {
        ResTblIterator {
            table: self,
            index: 0,
        }
    }

    #[inline(always)]
    fn name_table_offset(&self) -> usize {
        Header::FULL_SIZE + self.header.crc_table_count as usize * size_of::<HashEntry>()
    }

    #[inline(always)]
    fn hash_table_index(&self, index: usize) -> Option<HashTableIndex> {
        (index < self.header.crc_table_count as usize).then_some(HashTableIndex(index))
    }
}

#[cfg(feature = "alloc")]
impl super::ResourceSizeTable {
    /// Parse an owned table from binary form
    pub fn from_binary(data: impl AsRef<[u8]>) -> Result<Self> {
        fn inner(data: &[u8]) -> Result<super::ResourceSizeTable> {
            let parser = ResTblReader::new(data)?;
            let mut crc_table = alloc::collections::BTreeMap::new();
            let mut name_table = alloc::collections::BTreeMap::new();
            for entry in parser.iter() {
                match entry {
                    TableEntry::Hash(entry) => crc_table.insert(entry.hash(), entry.value()),
                    TableEntry::Name(entry) => name_table.insert(entry.name(), entry.value()),
                };
            }
            Ok(super::ResourceSizeTable {
                crc_table,
                name_table,
            })
        }
        inner(data.as_ref())
    }

    /// Write the table in its binary format to bytes.
    pub fn to_binary(&self) -> alloc::vec::Vec<u8> {
        let size = Header::FULL_SIZE
            + size_of::<HashEntry>() * self.crc_table.len()
            + size_of::<NameEntry>() * self.name_table.len();
        let mut buffer = alloc::vec![0u8; size];
        Header {
            version: 1,
            string_block_size: size_of::<Name>() as u32,
            crc_table_count: self.crc_table.len() as u32,
            name_table_count: self.name_table.len() as u32,
        }
        .write(&mut buffer);
        let mut pos = Header::FULL_SIZE;
        for (hash, value) in &self.crc_table {
            HashEntry {
                hash: *hash,
                value: *value,
            }
            .write(&mut buffer[pos..]);
            pos += size_of::<HashEntry>();
        }
        for (name, value) in &self.name_table {
            NameEntry {
                name: *name,
                value: *value,
            }
            .write(&mut buffer[pos..]);
            pos += size_of::<NameEntry>();
        }
        buffer
    }
}

#[cfg(test)]
mod test {
    use crate::test::DATA;

    #[test]
    fn parse() {
        let parser = super::ResTblReader::new(DATA).unwrap();
        dbg!(parser.header());
        #[cfg(feature = "alloc")]
        {
            let table = crate::ResourceSizeTable::from_parser(&parser);
            assert_eq!(
                parser.header.crc_table_count as usize,
                table.crc_table.len(),
            );
            assert_eq!(
                parser.header.name_table_count as usize,
                table.name_table.len(),
            );
        }
    }

    #[test]
    fn lookup() {
        let parser = super::ResTblReader::new(DATA).unwrap();
        let entry = parser
            .get("Cooking/CookingTable.game__cooking__Table.bgyml")
            .unwrap();
        dbg!(entry);
        assert!(parser.get("Pack/Actor/Nonexistent.pack").is_none());
    }

    #[cfg(feature = "alloc")]
    #[test]
    fn serialize() {
        let table = crate::ResourceSizeTable::from_binary(DATA).unwrap();
        let bytes = table.to_binary();
        let table2 = crate::ResourceSizeTable::from_binary(&bytes).unwrap();
        assert_eq!(table.crc_table.len(), table2.crc_table.len());
        for (entry1, entry2) in table.crc_table.iter().zip(table2.crc_table.iter()) {
            assert_eq!(entry1, entry2);
        }
        assert_eq!(table.name_table.len(), table2.name_table.len());
        for (entry1, entry2) in table.name_table.iter().zip(table2.name_table.iter()) {
            assert_eq!(entry1, entry2);
        }
        assert_eq!(DATA, bytes);
    }
}


src/util.rs
#[cfg_attr(feature = "serde", derive(serde::Deserialize))]
#[cfg_attr(feature = "serde", serde(from = "&str"))]
#[derive(Clone, Copy)]
pub struct Name {
    inner: [u8; 160],
}

impl core::ops::Deref for Name {
    type Target = str;
    fn deref(&self) -> &Self::Target {
        self.as_str()
    }
}

impl core::fmt::Debug for Name {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        f.debug_struct("String")
            .field("length", &self.as_str().len())
            .field("value", &self.as_str())
            .finish()
    }
}

impl core::fmt::Display for Name {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        self.as_str().fmt(f)
    }
}

impl PartialEq for Name {
    fn eq(&self, other: &Self) -> bool {
        self.as_str() == other.as_str()
    }
}

impl PartialEq<&str> for Name {
    fn eq(&self, other: &&str) -> bool {
        self.as_str() == *other
    }
}

impl PartialEq<&str> for &Name {
    fn eq(&self, other: &&str) -> bool {
        self.as_str() == *other
    }
}

impl Eq for Name {}

impl PartialOrd for Name {
    fn partial_cmp(&self, other: &Self) -> Option<core::cmp::Ordering> {
        self.as_str().partial_cmp(other.as_str())
    }
}

impl Ord for Name {
    fn cmp(&self, other: &Self) -> core::cmp::Ordering {
        self.as_str().cmp(other.as_str())
    }
}

impl PartialOrd<&str> for Name {
    fn partial_cmp(&self, other: &&str) -> Option<core::cmp::Ordering> {
        self.as_str().partial_cmp(*other)
    }
}

impl PartialOrd<&str> for &Name {
    fn partial_cmp(&self, other: &&str) -> Option<core::cmp::Ordering> {
        self.as_str().partial_cmp(*other)
    }
}

impl Name {
    #[inline(always)]
    pub fn as_str(&self) -> &str {
        unsafe {
            let zero_idx = self.inner.iter().position(|c| *c == 0).unwrap_unchecked();
            core::str::from_utf8_unchecked(&self.inner[..zero_idx])
        }
    }
}

impl AsRef<str> for Name {
    #[inline(always)]
    fn as_ref(&self) -> &str {
        self.as_str()
    }
}

impl From<&str> for Name {
    fn from(value: &str) -> Self {
        let mut inner: [u8; 160] = unsafe { core::mem::zeroed() };
        let len = value.len().min(160);
        inner[..len].copy_from_slice(value.as_bytes());
        Self { inner }
    }
}

impl TryFrom<&[u8]> for Name {
    type Error = crate::Error;

    fn try_from(value: &[u8]) -> Result<Self, Self::Error> {
        let mut inner: [u8; 160] = unsafe { core::mem::zeroed() };
        let mut len = 0;
        for (dest, src) in inner.iter_mut().zip(value.iter()) {
            if *src != 0 {
                *dest = *src;
                len += 1;
            } else {
                break;
            }
        }
        Ok(core::str::from_utf8(&inner[..len]).map(|_| Self { inner })?)
    }
}

#[cfg(feature = "serde")]
impl serde::Serialize for Name {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        self.as_str().serialize(serializer)
    }
}

pub(crate) fn read_u32(value: &[u8], offset: Option<usize>) -> crate::Result<u32> {
    let offset = offset.unwrap_or_default();
    if value.len() < 4 + offset {
        Err(crate::Error::InsufficientData(
            value.len() - offset,
            "4 bytes for u32",
        ))
    } else {
        Ok(u32::from_le_bytes(unsafe {
            value[offset..offset + 4].try_into().unwrap_unchecked()
        }))
    }
}

/// CRC hash function matching that used in BOTW/TOTK.
#[inline]
pub const fn hash_name(name: &str) -> u32 {
    let mut crc = 0xFFFFFFFF;
    let mut i = 0;
    while i < name.len() {
        crc ^= name.as_bytes()[i] as u32;
        let mut j = 0;
        while j < 8 {
            if crc & 1 == 1 {
                crc = (crc >> 1) ^ 0xEDB88320;
            } else {
                crc >>= 1;
            }
            j += 1;
        }
        i += 1;
    }
    !crc
}


src/lib.rs
//! # restbl
//!
//! A simple library to handle RSTB/RESTBL (resource size table) files from *The
//! Legend of Zelda: Tears of the Kingdom*. Features:
//! - Quick, zero-allocation parser
//! - Optional `alloc` feature to support editable table which can be serialized to
//!   binary or (with the `yaml` feature) YAML.
//! - `no_std` support (optional `std` feature)
//! - optional Serde support (`serde` feature)
//! - `aarch64-nintendo-switch-freestanding` support (without the `std` feature)
//!
//! ## Example Usage
//!
//! ```rust
//! use restbl::bin::ResTblReader;
//!
//! let bytes = std::fs::read("test/ResourceSizeTable.Product.110.rsizetable").unwrap();
//!
//! // Setup the quick, zero-allocation reader
//! let reader = ResTblReader::new(bytes.as_slice()).unwrap();
//! // Lookup an RSTB value
//! assert_eq!(
//!     reader.get("Bake/Scene/MainField_G_26_43.bkres"),
//!     Some(31880)
//! );
//!
//! #[cfg(feature = "alloc")]
//! {
//!     use restbl::ResourceSizeTable;
//!     // Parse RSTB into owned table
//!     let mut table = ResourceSizeTable::from_parser(&reader);
//!     // Set the size for a resource
//!     table.set("TexToGo/Etc_BaseCampWallWood_A_Alb.txtg", 777);
//!     // Check the size
//!     assert_eq!(
//!         table.get("TexToGo/Etc_BaseCampWallWood_A_Alb.txtg"),
//!         Some(777)
//!     );
//!     // Dump to YAML, if `yaml` feature enabled
//!     #[cfg(feature = "yaml")]
//!     {
//!         let json_table = table.to_text();
//!         // From YAML back to RSTB
//!         let new_table = ResourceSizeTable::from_text(&json_table).unwrap();
//!     }
//! }
//! ```
//!
//! ## Building for Switch
//!
//! To build for Switch, you will need to use the
//! `aarch64-nintendo-switch-freestanding` target. The `std` feature is not
//! supported, so you will need to use `--no-default-features`. Since [`cargo
//! nx`](https://github.com/aarch64-switch-rs/cargo-nx) does not seem to support
//! passing feature flags, you will need to run the full command yourself, as
//! follows:
//!
//! ```
//! cargo build -Z build-std=core,compiler_builtins,alloc --target aarch64-nintendo-switch-freestanding --no-default-features
//! ```
//!
//! ## License
//!
//! This software is licensed under the terms of the GNU General Public License,
//! version 3 or later.
#![cfg_attr(not(any(feature = "std", test)), no_std)]
#[cfg(feature = "alloc")]
extern crate alloc;
pub mod bin;
#[cfg(feature = "yaml")]
mod text;
mod util;

#[cfg(feature = "alloc")]
use alloc::{
    borrow::{Cow, ToOwned},
    collections::BTreeMap,
};
#[cfg(not(feature = "alloc"))]
pub use bin::ResTblReader;
use thiserror_no_std::Error;
use util::Name;

/// Result type for this create
pub type Result<T> = core::result::Result<T, Error>;

/// Error type for this crate
#[derive(Debug, Error)]
pub enum Error {
    #[error("Insufficient data: found {0} bytes, expected {1}")]
    InsufficientData(usize, &'static str),
    #[error("Invalid magic: {0:?}, expected \"RESTBL\"")]
    InvalidMagic([u8; 6]),
    #[error("Invalid table size: {0}, expected {1}")]
    InvalidTableSize(usize, usize),
    #[error(transparent)]
    Utf8Error(#[from] core::str::Utf8Error),
    #[error("Buffer too small for output: found {0} bytes, requires at least {1}")]
    InsufficientBuffer(usize, usize),
    #[cfg(feature = "std")]
    #[error(transparent)]
    IoError(#[from] std::io::Error),
    #[cfg(all(feature = "alloc", feature = "yaml"))]
    #[error("Invalid YAML line: {0}")]
    YamlError(alloc::string::String),
    #[cfg(feature = "yaml")]
    #[error("Invalid number in YAML line: {0}")]
    YamlInvalidNumber(#[from] core::num::ParseIntError),
}

/// Represents an index into the RSTB, which can be a canonical resource path or
/// its hash
#[derive(Debug)]
pub enum TableIndex<'a> {
    HashIndex(u32),
    #[cfg(feature = "alloc")]
    StringIndex(Cow<'a, str>),
    #[cfg(not(feature = "alloc"))]
    StringIndex(&'a str),
}

impl From<u32> for TableIndex<'_> {
    fn from(value: u32) -> Self {
        TableIndex::HashIndex(value)
    }
}

impl<'a> From<&'a str> for TableIndex<'a> {
    fn from(value: &'a str) -> Self {
        #[cfg(feature = "alloc")]
        {
            TableIndex::StringIndex(value.into())
        }
        #[cfg(not(feature = "alloc"))]
        {
            TableIndex::StringIndex(value)
        }
    }
}

impl<'a> From<&'a Name> for TableIndex<'a> {
    fn from(value: &'a Name) -> Self {
        #[cfg(feature = "alloc")]
        {
            TableIndex::StringIndex(Cow::Borrowed(value.as_str()))
        }
        #[cfg(not(feature = "alloc"))]
        {
            TableIndex::StringIndex(value.as_str())
        }
    }
}

#[cfg(feature = "alloc")]
impl From<Name> for TableIndex<'_> {
    fn from(value: Name) -> Self {
        TableIndex::StringIndex(Cow::Owned(value.as_str().to_owned()))
    }
}

#[cfg(feature = "alloc")]
impl From<alloc::string::String> for TableIndex<'_> {
    fn from(value: alloc::string::String) -> Self {
        TableIndex::StringIndex(value.into())
    }
}

/// Data structure representing Tears of the Kingdom's resource size table
/// (`ResourceSizeTable.Product.rsizetable.zs`). Requires the `alloc` feature.
/// Can be serialized or deserialized to binary or (with the `text` feature) a
/// YAML document.
#[cfg(feature = "alloc")]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
#[derive(Debug, Default, Clone, PartialEq)]
pub struct ResourceSizeTable {
    pub crc_table: BTreeMap<u32, u32>,
    pub name_table: BTreeMap<Name, u32>,
}

#[cfg(feature = "alloc")]
impl ResourceSizeTable {
    /// Construct an empty table
    pub fn new() -> Self {
        Self::default()
    }

    /// Construct an owned table from a fast readonly parser
    pub fn from_parser(parser: &bin::ResTblReader<'_>) -> Self {
        let mut crc_table = BTreeMap::new();
        let mut name_table = BTreeMap::new();
        for entry in parser.iter() {
            match entry {
                bin::TableEntry::Hash(entry) => crc_table.insert(entry.hash(), entry.value()),
                bin::TableEntry::Name(entry) => name_table.insert(entry.name(), entry.value()),
            };
        }
        ResourceSizeTable {
            crc_table,
            name_table,
        }
    }

    /// Get the total number of hash and name entries in the table
    #[inline(always)]
    pub fn len(&self) -> usize {
        self.crc_table.len() + self.name_table.len()
    }

    #[inline(always)]
    pub fn is_empty(&self) -> bool {
        self.len() == 0
    }

    /// Check if the specified hash or resource name is present in the table.
    /// Checks the name table first (if applicable) and then the hash table.
    pub fn contains<'i, I: Into<TableIndex<'i>>>(&self, needle: I) -> bool {
        fn inner(tbl: &ResourceSizeTable, needle: TableIndex) -> bool {
            match needle {
                TableIndex::HashIndex(hash) => tbl.crc_table.contains_key(&hash),
                TableIndex::StringIndex(name) => {
                    tbl.name_table.contains_key(&Name::from(name.as_ref())) || {
                        let hash = util::hash_name(&name);
                        tbl.crc_table.contains_key(&hash)
                    }
                }
            }
        }
        inner(self, needle.into())
    }

    /// Returns the RSTB value for the specified hash or resource name if
    /// present. Checks the name table first (if applicable) and then the hash
    /// table.
    pub fn get<'i, I: Into<TableIndex<'i>>>(&self, needle: I) -> Option<u32> {
        fn inner(tbl: &ResourceSizeTable, needle: TableIndex) -> Option<u32> {
            match needle {
                TableIndex::HashIndex(hash) => tbl.crc_table.get(&hash),
                TableIndex::StringIndex(name) => {
                    tbl.name_table.get(&Name::from(name.as_ref())).or_else(|| {
                        let hash = util::hash_name(&name);
                        tbl.crc_table.get(&hash)
                    })
                }
            }
            .copied()
        }
        inner(self, needle.into())
    }

    /// Returns a mutable reference to the RSTB value for the specified hash or
    /// resource name if present. Checks the name table first (if applicable)
    /// and then the hash table.
    pub fn get_mut<'i, I: Into<TableIndex<'i>>>(&mut self, needle: I) -> Option<&mut u32> {
        fn inner<'a>(
            tbl: &'a mut ResourceSizeTable,
            needle: TableIndex<'_>,
        ) -> Option<&'a mut u32> {
            match needle {
                TableIndex::HashIndex(hash) => tbl.crc_table.get_mut(&hash),
                TableIndex::StringIndex(name) => tbl
                    .name_table
                    .get_mut(&Name::from(name.as_ref()))
                    .or_else(|| {
                        let hash = util::hash_name(&name);
                        tbl.crc_table.get_mut(&hash)
                    }),
            }
        }
        inner(self, needle.into())
    }

    /// Set the RSTB value for the specified hash or resource name, returning
    /// the original value if present. Checks the name table first (if
    /// applicable) and then the hash table.
    pub fn set<'i, I: Into<TableIndex<'i>>>(&mut self, res: I, value: u32) -> Option<u32> {
        fn inner(tbl: &mut ResourceSizeTable, needle: TableIndex, value: u32) -> Option<u32> {
            match needle {
                TableIndex::HashIndex(hash) => tbl.crc_table.insert(hash, value),
                TableIndex::StringIndex(name) => {
                    match tbl.name_table.entry(Name::from(name.as_ref())) {
                        alloc::collections::btree_map::Entry::Occupied(mut e) => {
                            Some(e.insert(value))
                        }
                        alloc::collections::btree_map::Entry::Vacant(_) => {
                            let hash = util::hash_name(&name);
                            tbl.crc_table.insert(hash, value)
                        }
                    }
                }
            }
        }
        inner(self, res.into(), value)
    }

    /// Remove the RSTB value for the specified hash or resource name, returning
    /// the original value if present. Checks the name table first (if
    /// applicable) and then the hash table.
    pub fn remove<'i, I: Into<TableIndex<'i>>>(&mut self, res: I) -> Option<u32> {
        fn inner(tbl: &mut ResourceSizeTable, needle: TableIndex<'_>) -> Option<u32> {
            match needle {
                TableIndex::HashIndex(hash) => tbl.crc_table.remove(&hash),
                TableIndex::StringIndex(name) => tbl
                    .name_table
                    .remove(&Name::from(name.as_ref()))
                    .or_else(|| {
                        let hash = util::hash_name(&name);
                        tbl.crc_table.remove(&hash)
                    }),
            }
        }
        inner(self, res.into())
    }

    /// Set multiple RSTB entries from an iterator
    pub fn extend<'i, N: Into<TableIndex<'i>>, I: Iterator<Item = (N, u32)>>(&mut self, iter: I) {
        fn inner<'i, I: Iterator<Item = (TableIndex<'i>, u32)>>(
            tbl: &mut ResourceSizeTable,
            iter: I,
        ) {
            for (k, v) in iter {
                tbl.set(k, v);
            }
        }
        inner(self, iter.map(|(k, v)| (k.into(), v)))
    }
}

#[cfg(test)]
mod test {
    pub(crate) static DATA: &[u8] =
        include_bytes!("../test/ResourceSizeTable.Product.110.rsizetable");
}
